activity/manager.go

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package activity
+
+import (
+	"errors"
+	"github.com/eclipse/che-machine-exec/exec"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"os"
+	"time"
+)
+
+var (
+	WorkspaceAPIResource = &metav1.APIResource{
+		Name:       "workspaces",
+		Group:      "workspace.che.eclipse.org",
+		Version:    "v1alpha1",
+		Namespaced: true,
+	}
+
+	WorkspaceGroupVersion = &schema.GroupVersion{
+		Group:   "workspace.che.eclipse.org",
+		Version: "v1alpha1",
+	}
+)
+
+type Manager interface {
+	//Start starts users activity and scheduling workspace stoppping if there is no activity for idle timeout
+	Start()
+
+	//Tick registers users activity and postpones workpace stopping by inactivity
+	Tick()
+}
+
+func New(idleTimeout, stopRetryPeriod time.Duration) (Manager, error) {
+	if idleTimeout < 0 {
+		return &noOpManager{}, nil
+	}
+
+	if stopRetryPeriod <= 0 {
+		return nil, errors.New("stop retry period must be greater than 0")
+	}
+
+	namespace := exec.GetNamespace()
+	if namespace == "" {
+		return nil, errors.New("unable to evaluate the current namespace that is needed for activity manager works correctly")
+	}
+
+	workspaceName, isFound := os.LookupEnv("CHE_WORKSPACE_NAME")
+	if !isFound {
+		return nil, errors.New("CHE_WORKSPACE_NAME env must be set for activity manager works correctly")
+	}
+
+	return managerImpl{
+		namespace:       namespace,
+		workspaceName:   workspaceName,
+		idleTimeout:     idleTimeout,
+		stopRetryPeriod: stopRetryPeriod,
+	}, nil
+}
+
+// noOpManager should be used if idle timeout is configured less 0
+// invocation its method does not have affect
+type noOpManager struct{}
+
+func (m noOpManager) Tick()  {}
+func (m noOpManager) Start() {}
+
+type managerImpl struct {
+	namespace     string
+	workspaceName string
+
+	idleTimeout     time.Duration
+	stopRetryPeriod time.Duration
+
+	activityC chan bool
+}
+
+func (m managerImpl) Tick() {
+	m.activityC <- true
+}
+
+func (m managerImpl) Start() {
+	if m.activityC != nil {
+		//it's already started
+		return
+	}
+	logrus.Infof("Activity tracker is run and workspace will be stopped in %s if there is no activity", m.idleTimeout)
+	m.activityC = make(chan bool)
+	timer := time.NewTimer(m.idleTimeout)
+	select {
+	case <-timer.C:
+		if err := m.stopWorkspace(); err != nil {
+			timer.Reset(m.stopRetryPeriod)
+			logrus.Errorf("Failed to stop workspace. Will retry in %s. Cause: %s", m.stopRetryPeriod, err)
+		} else {
+			logrus.Info("Workspace is successfully stopped by inactivity. Bye")
+			return
+		}
+	case <-m.activityC:
+		logrus.Debug("Activity is reported. Resetting timer")
+		if !timer.Stop() {
+			<-timer.C
+		}
+		timer.Reset(m.idleTimeout)
+	}
```

Comment:
This needs to be done in a goroutine, otherwise it blocks startup.

Reply:
Makes sense. I wonder why it worked in my local testing :exploding_head: